Add support for more mocha functions

felixSchl opened the pull request:
This patch critically refactors `Mocha.purs` and brings a range of new features:
- Add prime version for every hook and `it` to add a custom timeout
- Add `xit`, `xdescribe` to mark pending tests
- Add all variants of `it` and `xit` for both the sync and the async version
- Add utility methods to make changes easier (`apply0 ... 2` and `syncImpl` and `asyncImpl`). This should make it easier to add more of mochas different suit-styles later on.

To get a better idea, you can contrast the exports of before and after the patch.
This patch **should not** break any existing code, however it does remove `DoIt(..)` and `DoDescribe(..)` from the exports which from what I gather only had to be exported to make the compiler happy. The new version of the code doesn't need the internal aliases and hence does not have to export them.

Also, the code is tested only manually as we have no test suite set up as it stands.

felixSchl committed:
Refactor most of `Mocha.purs`

o Add timeout via prime versions, i.e:
  The after hook `after` now has a variant `after'` that takes a number
  as it's first argument
o Add `xit`, `xdescribe` to mark pending tests
o Add all variants of `it` for both the sync and the async version
o Add utility methods to make changes easier (`apply0 ... 2` and
  `syncImpl` and `asyncImpl`). This should make it easier to add more of
  mochas different suit-styles later on.

Fresheyeball commented:
Wow, this is a really big shift, and introduces some new concepts. I will read fully later, but for starts can all the exported functions have type signatures?

felixSchl commented:
Yeah sure, for now have i pinned down each of `it`, `xit`, ... 's basetypes to ensure they run with the right effects. But can I can explicitly ad annotations for all of them.

Fresheyeball commented:
I would like explicit type annotations for all exported functions please. Also `xIt` is not every descriptive, can you explain to me what it does?

felixSchl committed:
Add explicit type annotations for exported types

felixSchl commented:
I only learned about `xit` when reading the mocha sources to write this code, doesn't seem to be wildly used. However, an `it` without a callback does the same, it creates a "pending" test, i.e. a test to be written. I guess we can remove the permutations of `xit` and maybe use the name for `xit :: forall eff. String -> Eff eff Unit` to mark a not-implemented test? To "comment out" an existing test, one would just use the skip version of it.

felixSchl commented:
Another thought that crossed my mind is that we are changing the mocha interface slightly by making timeout a parameter. Not only adds it code bloat to the library, but there are *probably* reasons that the mocha guys decided not to. Maybe we need a more beefed up handler type that allows do these sort of modifications inside a `do` block?

felixSchl committed:
Remove `xit` and `describe` variations

o Remove the async/timeout/only/skip variations of `xit`
o Remove the only/skip variations of `describe`
o Add `xdescribe'` which takes no callback

felixSchl commented:
I pushed another commit trying out specialising `xit` and `xdescribe`. I think I will leave it at that until further feedback.

Fresheyeball commented:
ok, so if `xit` is equivalent to `it.skip` then lets remove the `xIt` and `xdescribe` from the bindings completely. If consumers of this lib want to skip a test, `itSkip` is sufficient. Just my feeling, but redundancy in an api drives me nuts.

Fresheyeball commented:
Regarding timeouts as a parameter, it might be necessary to access the functionality of `this.timeout` from PureScript. However, having most tests use `this.timeout(0);` also effect the running of that test. I'm starting to thing that a version of `it` using `this.timeout` should have separate ffi code from the normal version of `it` that does not use `this.timeout`, to ensure an intuitive separation in behavior. 

Think through this further, I think that `it'` is not sufficient to express the timeout behavior. Consider the following api:

```
it 
itAsync 
itTimed -- version with `this.timeout`
itAsyncTimed -- version with `this.timeout`
itSkip
```

felixSchl commented:
> However, having most tests use `this.timeout(0);` also effect the running of that test. 

Not necessarily because timeout is only applied if `timeout > 0`. I did this to avoid using `Maybe`, but we could use that instead. In terms of naming, sure. It's just that these are really the only two overloads of the function and keeps the test suites leaner. What I mean is that there's not a significant difference between `itAsync` and `itAsync'` that it would warrant it's own name (just my opinion) because all `itAsync` have a default timeout anyway (so they are all "timed").

felixSchl commented:
> ok, so if `xit` is equivalent to `it.skip` then lets remove the `xIt` and `xdescribe` from the bindings completely.

Sure, we may want a noop function though, since those functions require callbacks

felixSchl commented:
@Fresheyeball, what is required for these changes to make it into master? I am in need of a testing framework and would like to stick with mocha, but I need the added functionality as introduced in this PR.

Fresheyeball commented:
I am abandoning this repo (purescript-spec seems to have won the testing). @felixSchl would you like to take over?

felixSchl commented:
I won't have the time to maintain this, I am currently really only on and off with purescript. purescript-spec looks pretty promising.

Fresheyeball commented:
Well I've been glad for your help regardless. Looks like this one is dead for now.
